IAS authentications flows (#2089)
cap-js/docs: java/security.md

:::
The mock user `Alice` is assigned to the mock tenant `CrazyCars` for which the features `cruise` and `park` are enabled.

## Connecting to IAS Services { #outbound-auth }

CAP Java supports the consumption of IAS-based services of various kinds:

* [Internal Services](#internal-app) bound to the same IAS instance of the provider application.
* [External IAS](#app-to-app) applications consumed by providing a destination.
* [BTP reuse services](#ias-reuse) consumed via service binding.

![The TAM graphic is explained in the accompanying text.](./assets/java-ias.png){width="800px" }

Regardless of the kind of service, CAP provides a unified integration as Remote Service as described in the [documentation](/java/cqn-services/remote-services#remote-odata-services).
Basic communication setup and user propagation is addressed under the hood, for example, an mTLS handshake is performed in case of service-2-service communication.

> ### Internal Services (same IAS) {#internal-app}

Reviewer: I'll hold off on approving the PR as this section is empty.

Reviewer 2: ```suggestion
```

The CAP Java application as an IAS server needs to:

1. Configure [IAS authentication](/java/security#xsuaa-ias).
2. Expose an API in the IAS service instance.

    ::: details Sample IAS instance of server

    ```yaml
    - name: server-identity
        type: org.cloudfoundry.managed-service
        parameters:
          service: identity
          service-plan: application
          config:
            multi-tenant: true
            provided-apis:
              - name: "review-api"
    ```

    :::

3. Prepare a CDS service endpoint for the exposed API.

    ::: details Sample CDS Service for the API

    ```cds
    service ReviewService @(requires: 'review-api') {
      [...]
    }
    ```

    :::

::: tip
The API identifiers exposed by the IAS instance in list `provided-apis` are granted as CAP roles after successful authentication.
:::

::: warning 
Use different CAP roles for technical clients without user propagation and for named business users.

Instead of using the same role, expose dedicated CDS services to technical clients which aren't accessible to business users and vice verse.
:::

> To set up a connection to such an IAS service, the client requires to do:

Reviewer 2: ```suggestion
To set up a connection to such an IAS service, the _consumer application_ requires to do:
```

The CAP reuse service (provider) needs to:

1. Configure [IAS authentication](/java/security#xsuaa-ias).
2. Bind an IAS instance that exposes services and service plans.

    ::: details Sample IAS instance for provider

    ```yaml
    - name: server-identity
        type: org.cloudfoundry.managed-service
        parameters:
          service: identity
          service-plan: application
          config:
            multi-tenant: true
            catalog:
              services:
                - id: "1d5c23ee-1ce6-6130-4af4-26461bc6ef79"
                  name: "review-service"
                  plans:
                    - id: "2d5c23ee-1ce6-6130-4af4-26461bc6ef78"
                      name: "review-api"
    ```

    :::

3. Prepare a CDS service endpoint for the exposed API.

    ::: details Sample CDS Service for the API

    ```cds
    service ReviewService @(requires: 'review-api') {
      [...]
    }
    ```

    :::

The CAP consumer application (client) needs to:

1. Create and bind the provided service from the marketplace.

    ::: details Create and bind service instance.
    ```sh
    cf create-service review-service review-api review-service-instance

>     cf bind-service review-service-instance --binding-name review-service-binding

Reviewer 2: ```suggestion
    cf bind-service <your-app-name> review-service-instance --binding-name review-service-binding
```

Reviewer 2: <img width="1459" height="103" alt="image" src="https://github.com/user-attachments/assets/a4bafefe-39a9-4d24-8b02-751eb7850fe6" />

    :::

4. Use CQN queries to consume the reuse service (optional)

[Learn more about simplified Remote Service configuration with bindings](/java/cqn-services/remote-services#service-binding-based-scenarios) {.learn-more}

::: tip Service plan name as CAP role
The service plan names as specified in `consumed-services` in the IAS instance are granted as CAP roles after successful authentication.
:::

::: warning  Use different roles for technical and business users
Use different CAP roles for technical clients without user propagation and for named business users.

Instead of using the same role, expose dedicated CDS services to technical clients which aren't accessible to business users and vice versa.
:::

#### How to Authorize Callbacks

For bidirectional communication, callbacks from the reuse service to the CAP service need to be authorized as well.
Currently, there is no standadized way to achieve this in CAP so that custom codeing is required.
As a prerequisite*, the CAP service needs to know the clientId of the reuse service's IAS application which should be part of the binding exposed to the CAP service.

::: details Sample Code for Authorization of Callbacks

```java

> private void authorizeCallback() {
> 		UserInfo userInfo = runtime.getProvidedUserInfo();
> 		String azp = (String) userInfo.getAdditionalAttributes().get("azp");
> 		if(!userInfo.isSystemUser() || azp == null || !azp.equals(clientId)) {
> 			throw new ErrorStatusException(ErrorStatuses.FORBIDDEN);
> 		}
> 	}

Reviewer 2: ```suggestion
@Component
public class CallbackAuthorizationHandler {

  @Value("${reuse-service.client-id}")
  private String expectedClientId;

  private void authorizeCallback() {
    UserInfo userInfo = runtime.getProvidedUserInfo();
    String azp = (String) userInfo.getAdditionalAttributes().get("azp");

    if (!userInfo.isSystemUser() || azp == null || !azp.equals(expectedClientId)) {
      throw new ErrorStatusException(ErrorStatuses.FORBIDDEN);
```

Reviewer 2: Does that make sense? AI proposed that

Author: I'd rather stick to the original code snippet

For connection setup, it makes use of [IAS App-2-App flows](https://help.sap.com/docs/cloud-identity-services/cloud-identity-services/consume-apis-from-other-applications).

The CAP Java application as IAS server needs to

1. Configure [IAS authentication](/java/security#xsuaa-ias).
2. Expose an API in the IAS service instance.

    ::: details Sample IAS instance of server

    ```yaml
    - name: server-identity
        type: org.cloudfoundry.managed-service
        parameters:
          service: identity
          service-plan: application
          config:
            multi-tenant: true
            provided-apis:
              - name: "review-api"
    ```

    :::

3. Prepare a CDS service endpoint for the exposed API.

    ::: details Sample CDS Service for the API

    ```cds
    service ReviewService @(requires: 'review-api') {
      [...]
    }
    ```

    :::

::: tip
The API identifiers exposed by the IAS instance in list `provided-apis` are granted as CAP roles after successfull authentication.
:::

::: warning 
Avoid mixing CAP roles for technical clients without user propagation on the one hand and named business users on the other hand.

> Instead, expose dedicated CDS services to technical clients which are not accessible to business users and vica verse.

Reviewer: ```suggestion
Instead, expose dedicated CDS services to technical clients which are not accessible to business users and vice verse.
```
